### PR TITLE
libuv: fix zero-length writes

### PR DESCRIPTION
**Pull request by cbay**
Although evloop does accept writes with bufcnt == 0, libuv explicitely doesn't and would [raise an assert](https://github.com/libuv/libuv/blob/v1.x/src/unix/stream.c#L1409).

I've run into this with requests having Content-Length: 0.

### CODE REVIEW

File: lib/common/socket/uv-binding.c.h [hunk 1]
```diff
@@ -197,7 +197,10 @@ void do_write(h2o_socket_t *_sock, h2o_iovec_t *bufs, size_t bufcnt, h2o_socket_
     assert(sock->super._cb.write == NULL);
     sock->super._cb.write = cb;
 
-    uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, (uv_buf_t *)bufs, (int)bufcnt, on_do_write_complete);
+    if (bufcnt > 0)
+        uv_write(&sock->stream._wreq, (uv_stream_t *)sock->handle, (uv_buf_t *)bufs, (int)bufcnt, on_do_write_complete);
+    else
+        on_do_write_complete(&sock->stream._wreq, 0);
```
Comment (kazuho) [inline lib/common/socket/uv-binding.c.h:203]: I think we have to invoke `on_do_write_complete` asynchronously. Callers of `h2o_socket_write` do not expect the completion callback to be invoked while `h2o_socket_write` is running.

To defer the invocation, I think the easiest way would be to add a `h2o_timer_t` instance to `st_h2o_uv_socket_t` and do `h2o_timer_link` with a timeout value of zero.

**Review by kazuho — COMMENTED** [review]
Thank you for spotting the issue.

**Review by kazuho — COMMENTED** [review]
Thank you for the changes.

I think we need to unlink the timer in `do_dispose_socket` function, as `h2o_socket_close` might get called between when `h2o_socket_write` is called with a zero-sized vector and when the new write completion callback gets called.

Otherwise the PR looks good!

**Comment by cbay** [comment]
> I think we need to unlink the timer in `do_dispose_socket` function

Indeed! Thanks.

**Comment by i110** [comment]
Thank you!

**Pull request merged**